src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,57 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    double diffOfTimeMSec = prevTime->msecsTo(myTime);
+    double diffOfTime = diffOfTimeMSec/3600000;
+    double avgCurrent = (myCurrent+prevCurrent)/2;
+
+    double changeInAmpHour = avgCurrent * diffOfTime;
+    ampHourTotal += changeInAmpHour;
+    //if(ampHourTotal > BATTERY_AMP_HOUR_CAPACITY){ampHourTotal = BATTERY_AMP_HOUR_CAPACITY;}
+    ampHourUsed = BATTERY_AMP_HOUR_CAPACITY- ampHourTotal;
+    return ampHourUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(myCurrent>=0)
+    {
+        return false;
+    }
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double ampHoursLeft;
+    if(isCharging())
+    {
+        ampHoursLeft = BATTERY_AMP_HOUR_CAPACITY - ampHourTotal;
+    }
+    else
+    {
+        ampHoursLeft = ampHourTotal;
+    }
+    int current = myCurrent;
+    if(myCurrent < 0)
+    {
+        current *= -1;
+    }
+    double msecLeft = (ampHoursLeft * 3600000) / current;
+    QTime QTimeLeft = QTime(0,0,0,0);
+    QTimeLeft = QTimeLeft.addMSecs(msecLeft);
```

Comment:
These two lines can be compressed into one line using, `QTime timeLeft = QTime(0, 0).addMsecs(msecLeft);`